expression/builtin_arithmetic.go

```
@@ -550,6 +554,134 @@ func (s *builtinArithmeticDivideDecimalSig) evalDecimal(row []types.Datum) (*typ
+	return sig.setSelf(sig), nil
+}
+
+type builtinArithmeticIntDivideIntUnsignedSig struct{ baseIntBuiltinFunc }
+type builtinArithmeticIntDivideIntSig struct{ baseIntBuiltinFunc }
+type builtinArithmeticIntDivideDecimalSig struct{ baseIntBuiltinFunc }
+
+func (s *builtinArithmeticIntDivideIntUnsignedSig) evalInt(row []types.Datum) (val int64, isNull bool, err error) {
+	sc := s.ctx.GetSessionVars().StmtCtx
+	a, isNull, err := s.args[0].EvalInt(row, sc)
+	if isNull || err != nil {
+		return 0, isNull, errors.Trace(err)
+	}
+
+	b, isNull, err := s.args[1].EvalInt(row, sc)
+	if isNull || err != nil {
+		return 0, isNull, errors.Trace(err)
+	}
+
+	if b == 0 {
+		return 0, true, nil
+	}
+
+	if a < 0 && -a >= b {
+		return 0, false, types.ErrOverflow.GenByArgs("BIGINT", fmt.Sprintf("(%d, %d)", a, b))
+	}
+
+	if a != 0 && b < 0 && -b <= a {
+		return 0, false, types.ErrOverflow.GenByArgs("BIGINT UNSIGNED", fmt.Sprintf("(%d, %d)", a, b))
+	}
+
+	return a / b, false, nil
+}
+
+func (s *builtinArithmeticIntDivideIntSig) evalInt(row []types.Datum) (val int64, isNull bool, err error) {
+	sc := s.ctx.GetSessionVars().StmtCtx
+	a, isNull, err := s.args[0].EvalInt(row, sc)
+	if isNull || err != nil {
+		return 0, isNull, errors.Trace(err)
+	}
+
+	b, isNull, err := s.args[1].EvalInt(row, sc)
+	if isNull || err != nil {
+		return 0, isNull, errors.Trace(err)
+	}
+
+	if b == 0 {
+		return 0, true, nil
+	}
+
+	ret, err := types.DivInt64(a, b)
+	return ret, false, err
+}
+
+func (s *builtinArithmeticIntDivideDecimalSig) evalInt(row []types.Datum) (int64, bool, error) {
+	sc := s.ctx.GetSessionVars().StmtCtx
+	a, isNull, err := s.args[0].EvalDecimal(row, sc)
+	if isNull || err != nil {
+		return 0, isNull, errors.Trace(err)
+	}
+
+	b, isNull, err := s.args[1].EvalDecimal(row, sc)
+	if isNull || err != nil {
+		return 0, isNull, errors.Trace(err)
+	}
+
+	c := &types.MyDecimal{}
+	err = types.DecimalDiv(a, b, c, types.DivFracIncr)
+	if err == types.ErrDivByZero {
+		return 0, true, nil
```

when divided by zero, we can just throw this error, we'll handle it in strict mode in the future, see #4373 for detail